src/app/icd/client/DefaultICDClientStorage.cpp

```diff
@@ -0,0 +1,436 @@
+{
+    TLV::TLVType arrayType;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Array, arrayType));
+    for (auto & clientInfo : clientInfoVector)
+    {
+        TLV::TLVType ICDClientInfoContainerType;
+        ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, ICDClientInfoContainerType));
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(ClientInfoTag::kPeerNodeId), clientInfo.peer_node.GetNodeId()));
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(ClientInfoTag::kFabricIndex), clientInfo.peer_node.GetFabricIndex()));
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(ClientInfoTag::kStartICDCounter), clientInfo.start_icd_counter));
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(ClientInfoTag::kOffset), clientInfo.offset));
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(ClientInfoTag::kMonitoredSubject), clientInfo.monitored_subject));
+        ByteSpan buf(clientInfo.mSharedKey.As<Crypto::Aes128KeyByteArray>());
+        ReturnErrorOnFailure(writer.Put(TLV::ContextTag(ClientInfoTag::kSharedKey), buf));
+        ReturnErrorOnFailure(writer.EndContainer(ICDClientInfoContainerType));
+    }
+    return writer.EndContainer(arrayType);
+}
+
+CHIP_ERROR DefaultICDClientStorage::StoreEntry(ICDClientInfo & clientInfo)
+{
+    std::vector<ICDClientInfo> clientInfoVector;
+    size_t clientInfoSize = MaxICDClientInfoSize();
+    ReturnErrorOnFailure(Load(clientInfo.peer_node.GetFabricIndex(), clientInfoVector, clientInfoSize));
+
+    for (auto it = clientInfoVector.begin(); it != clientInfoVector.end(); it++)
+    {
+        if (clientInfo.peer_node.GetNodeId() == it->peer_node.GetNodeId())
+        {
+            ReturnErrorOnFailure(UpdateCounter(clientInfo.peer_node.GetFabricIndex(), false /*increase*/));
+            clientInfoVector.erase(it);
+            break;
+        }
+    }
+
+    clientInfoVector.push_back(clientInfo);
+    size_t total = clientInfoSize * clientInfoVector.size() + kArrayOverHead;
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+    backingBuffer.Calloc(total);
+    ReturnErrorCodeIf(backingBuffer.Get() == nullptr, CHIP_ERROR_NO_MEMORY);
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), total);
+
+    ReturnErrorOnFailure(Save(writer, clientInfoVector));
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+    ReturnErrorOnFailure(mpClientInfoStore->SyncSetKeyValue(
+        DefaultStorageKeyAllocator::ICDClientInfoKey(clientInfo.peer_node.GetFabricIndex()).KeyName(), backingBuffer.Get(),
+        static_cast<uint16_t>(len)));
+
+    return UpdateCounter(clientInfo.peer_node.GetFabricIndex(), true /*increase*/);
+}
+
+CHIP_ERROR DefaultICDClientStorage::UpdateCounter(FabricIndex fabricIndex, bool increase)
```

Comment:
What counter is it? It's the number of clients for the fabricIndex.

Please rename to `UpdateEntryCountForFabric(FabricIndex fabricIndex, bool increase)` and then make it private. Add 2 new public methods:

`IncreaseEntryCountForFabric`/`DecreaseEntryCountForFabric` that populate the true/false properly. Code will be more fluent.

Can do as a follow-up.

```suggestion
CHIP_ERROR DefaultICDClientStorage::IncreaseCounterForFabric(FabricIndex fabricIndex, bool increase)
```

Reply:
updated, thanks